rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+     the _index_ instead of the alias.
+  3. Before starting a migration:
+    1. Write a new dummy "version lock" document to the `.kibana` index with a
+      `migrationVersion` set to the current version of Kibana. If an outdated
+      node is started up after a migration was started it will detect that
+      newer documents are present in the index and refuse to start up.
+    2. Set the outdated index to read-only. Since `.kibana` is never advanced,
+      it will be pointing to a read-only index which prevent writes from
+      6.8+ releases which are already online.
+</details>
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor upgrades.
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### 4.5.1 Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
+   1. Read a batch of outdated documents from `.kibana_n`.
+   2. Transform documents by applying the migration functions.
+   3. Update the document batch in the same index using optimistic concurrency
+      control. 
+   4. If a batch fails due to an update version mismatch, repeat the entire
+      process.
+4. Once all documents are up to date, the migration is complete and Kibana can
+   start serving traffic.
+
+Advantages:
+- Not duplicating all documents into a new index will speed up migrations and
+  reduce the downtime window. This will be especially important for the future
+  requirement to support > 10k or > 100k documents.
+- We can check the health of an existing index before starting the migration,
+  but we cannot detect what kind of failures might occur while creating a new
+  index. Whereas retrying migrations will eventually recover from the errors
+  in (3.3), re-using an index allows us to detect these problems before trying
+  and avoid errors like (3.3.1) altogether.
+- Single index to backup instead of “index pattern” that matches any
+  `.kibana_n`.
+- Simplifies Kibana system index Elasticsearch plugin since it needs to work
+  on one index per "tenant".
+- Can prevent data loss when mixed versions of Kibana are both attempting a
+  migration using optimistic concurrency control.
+
+Drawbacks:
+- Cannot make breaking mapping changes (even though it was possible we never
+  introduced a breaking mapping change during 7.x).
+- Rollback is only possible by restoring a snapshot which requires educating
+  users to ensure that they don't rely on `.kibana_n` indices as backups.
+  (Apart from the need to educate users, snapshot restores provide many
+  benefits).
```

Comment:
Should we add some failsafe CLI option here?

For example, in 8.0 we could have an interactive prompt on the CLI that asks "Have you made a snapshot prior to upgrade? (y/N)" and also provide a CLI option to skip this check for automated upgrades (eg `--skip-upgrade-snapshot-check`).

This could be a forcing function to make admins at least think about this before running an upgrade.

Reply:
Adding an accompanying config option might make it easier to bypass in some environments like a pre-built docker image.

Though I'm a bit hesitant about the impact this might have, it could increase downtime when Kibana gets upgraded and an administrator doesn't immediately realise this prompt.

Reply:
> Though I'm a bit hesitant about the impact this might have, it could increase downtime when Kibana gets upgraded and an administrator doesn't immediately realise this prompt.

At the very least, this change would need to be in a major upgrade. It's also not a common pattern and I agree it may create more problems than it solves (though data loss is very difficult to come back from).

Alternatively, maybe we should consider asking users to define a snapshot repository that Kibana should use for upgrade backups. If we have a snapshot repository in place, we can automatically trigger snapshot backups before upgrades. If there is none configured, we can at least emit a warning during server start up.

In the future, we could even walk the user through the process of setting up an SLM policy for Kibana.

Reply:
I like this kind of automation, this is something we could do on Cloud, but the Kibana nodes don't know which one first started the migration, so they all need to take a snapshot. Finding 3 upgrade backup snapshots could be confusing.

We could inspect all the snapshot repositories and ensure that there's a recent snapshot of all kibana's indexes, if not we can print a warning.

Reply:
> We could inspect all the snapshot repositories and ensure that there's a recent snapshot of all kibana's indexes, if not we can print a warning.

Sounds like a good option 👍. We could also have opt-in behavior for bailing upgrades if no recent snapshots have been taken. I know the implicit backup we had before wasn't really a feature, but I want to make sure we don't lead customers down a path that could lead to data loss without an escape hatch.